File: src/main/go/leetcode/lists/hot100/LeetCode494.go
package hot100

func findTargetSumWays(nums []int, target int) int {
	var count = 0
	var helper func([]int, int, int, int)

	helper = func(nums []int, target int, index int, sum int) {

		if index == len(nums) {
			if target == sum {
				count++
			}
		} else {
			helper(nums, target, index+1, sum+nums[index])
			helper(nums, target, index+1, sum-nums[index])
		}
		return
	}
	helper(nums, target, 0, 0)
	return count
}

File: java/src/leetcode/problems/LeetCode494.java
package leetcode.problems;

/**
 * @author wujingxinit@outlook.com
 * @date 2023/12/6 13:50
 * @see knowledge.algorithms.dp.backpack.lintcode.BackpackV
 */
public class LeetCode494 {

    public int findTargetSumWays(int[] nums, int target) {
        int sum = 0;
        for (int n : nums) {
            sum += n;
        }
        if ((target + sum) % 2 == 1 || target + sum < 0) {
            return 0;
        }
        int amount = (target + sum) / 2;
        int[] dp = new int[amount + 1];
        dp[0] = 1;
        for (int num : nums) {
            for (int j = amount; j >= num; j--) {
                dp[j] += dp[j - num];
            }
        }
        return dp[amount];
    }
}
